### PR TITLE
in_dummy: Add interval settings.

### PR DESCRIPTION
**Pull request by franciscovalentecastro**
This PR adds new `interval_sec` and `interval_nsec` settings to the `in_dummy` input plugin. This settings enable to set an arbitrary time interval in which to generate dummy logs. When `interval_sec/nsec` is set correctly, the `rate` configuration is ignored.
Documentation update in https://github.com/fluent/fluent-bit-docs/pull/1239.

### Configuration
```
[INPUT]
    Name          dummy
    Dummy         {"message": "custom dummy"}
    Interval_sec  1
    Interval_nsec 500000000
    Rate          0

[OUTPUT]
    Name   stdout
    Match  *
```

### Output Log
With this config a log message is generated every 1.5 seconds).

```
./fluent-bit -c fluent-bit.conf
Fluent Bit v2.2.0
* Copyright (C) 2015-2023 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/10/27 18:06:56] [ info] [fluent bit] version=2.2.0, commit=785f45a16d, pid=1533032
[2023/10/27 18:06:56] [ info] [storage] ver=1.5.1, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/10/27 18:06:56] [ info] [cmetrics] version=0.6.3
[2023/10/27 18:06:56] [ info] [ctraces ] version=0.3.1
[2023/10/27 18:06:56] [ info] [input:dummy:dummy.0] initializing
[2023/10/27 18:06:56] [ info] [input:dummy:dummy.0] storage_strategy='memory' (memory only)
[2023/10/27 18:06:56] [ info] [sp] stream processor started
[2023/10/27 18:06:56] [ info] [output:stdout:stdout.0] worker #0 started
[0] dummy.0: [[1698430017.644439396, {}], {"message"=>"custom dummy"}]
[0] dummy.0: [[1698430019.144370981, {}], {"message"=>"custom dummy"}]
[0] dummy.0: [[1698430020.644434995, {}], {"message"=>"custom dummy"}]
[0] dummy.0: [[1698430022.144357678, {}], {"message"=>"custom dummy"}]
^C[2023/10/27 18:07:03] [engine] caught signal (SIGINT)
[2023/10/27 18:07:03] [ warn] [engine] service will shutdown in max 5 seconds
[2023/10/27 18:07:03] [ info] [input] pausing dummy.0
[2023/10/27 18:07:03] [ info] [engine] service has stopped (0 pending tasks)
[2023/10/27 18:07:03] [ info] [input] pausing dummy.0
[2023/10/27 18:07:03] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2023/10/27 18:07:03] [ info] [output:stdout:stdout.0] thread worker #0 stopped
```

### Valgrind Output
```
valgrind --leak-check=yes ./fluent-bit   --config ./fluent-bit.conf
==1533023== Memcheck, a memory error detector
==1533023== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==1533023== Using Valgrind-3.16.1 and LibVEX; rerun with -h for copyright info
==1533023== Command: ./fluent-bit --config ./fluent-bit.conf
==1533023==
Fluent Bit v2.2.0
* Copyright (C) 2015-2023 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/10/27 18:06:01] [ info] [fluent bit] version=2.2.0, commit=785f45a16d, pid=1533023
[2023/10/27 18:06:01] [ info] [storage] ver=1.5.1, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/10/27 18:06:01] [ info] [cmetrics] version=0.6.3
[2023/10/27 18:06:01] [ info] [ctraces ] version=0.3.1
[2023/10/27 18:06:01] [ info] [input:dummy:dummy.0] initializing
[2023/10/27 18:06:01] [ info] [input:dummy:dummy.0] storage_strategy='memory' (memory only)
[2023/10/27 18:06:01] [ info] [sp] stream processor started
[2023/10/27 18:06:01] [ info] [output:stdout:stdout.0] worker #0 started
[0] dummy.0: [[1698429961.660134534, {}], {"message"=>"custom dummy"}]
[0] dummy.0: [[1698429963.144432977, {}], {"message"=>"custom dummy"}]
[0] dummy.0: [[1698429964.644634450, {}], {"message"=>"custom dummy"}]
[0] dummy.0: [[1698429966.144556969, {}], {"message"=>"custom dummy"}]
^C[2023/10/27 18:06:07] [engine] caught signal (SIGINT)
[2023/10/27 18:06:07] [ warn] [engine] service will shutdown in max 5 seconds
[2023/10/27 18:06:07] [ info] [input] pausing dummy.0
[2023/10/27 18:06:07] [ info] [engine] service has stopped (0 pending tasks)
[2023/10/27 18:06:07] [ info] [input] pausing dummy.0
[2023/10/27 18:06:07] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2023/10/27 18:06:07] [ info] [output:stdout:stdout.0] thread worker #0 stopped
==1533023==
==1533023== HEAP SUMMARY:
==1533023==     in use at exit: 0 bytes in 0 blocks
==1533023==   total heap usage: 1,716 allocs, 1,716 frees, 1,779,752 bytes allocated
==1533023==
==1533023== All heap blocks were freed -- no leaks are possible
==1533023==
==1533023== For lists of detected and suppressed errors, rerun with: -s
==1533023== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```
----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [x] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [ ] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**

- [x] Documentation required for this feature

**Backporting**

- [ ] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

File: plugins/in_dummy/in_dummy.c
Comment (franciscovalentecastro) [reply plugins/in_dummy/in_dummy.c]: Renamed to `tm_interval_fractional`.

File: plugins/in_dummy/in_dummy.c
Comment (franciscovalentecastro) [reply plugins/in_dummy/in_dummy.c]: Added `tm_interval_seconds` as an intermediate value.

**Review by braydonk — APPROVED** [review]

**Review by braydonk — COMMENTED** [review]
Realized after talking offline that we should add some tests for this to `in_simple_systems.c`

**Comment by franciscovalentecastro** [comment]
> Realized after talking offline that we should add some tests for this to `in_simple_systems.c`

Added unit tests to verify `rate` amount of logs were generated after one `time_interval` passed (plus 4 seconds of added time). This PR is ready for a re-review @braydonk .

**Review by braydonk — DISMISSED** [review]
LGTM, I'm ready to send it along to a maintainer.

**Comment by braydonk** [comment]
`CODEOWNERS` suggests we should ping @nokute78 for review here.

File: plugins/in_dummy/in_dummy.c
Comment (nokute78) [inline plugins/in_dummy/in_dummy.c:408]: How about setting the property name to `Interval_Sec` ?
Many plugins support `Interval_Sec` to set polling interval in second. It is consistent.

```
$ find . -name "*.c"|xargs grep -i \"interval_sec\"
./in_winevtlog/in_winevtlog.c:      FLB_CONFIG_MAP_TIME, "interval_sec", "1s",
./in_random/random.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_kubernetes_events/kubernetes_events.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_proc/in_proc.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_winlog/in_winlog.c:      FLB_CONFIG_MAP_TIME, "interval_sec", "1s",
./in_mem/mem.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_calyptia_fleet/in_calyptia_fleet.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_winstat/winstat.c:      FLB_CONFIG_MAP_TIME, "interval_sec", "1s",
./in_cpu/cpu.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_event_type/event_type.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_docker/docker.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_exec/in_exec.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_thermal/in_thermal.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_netif/in_netif.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_disk/in_disk.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_exec_wasi/in_exec_wasi.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_head/in_head.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
./in_health/health.c:      FLB_CONFIG_MAP_INT, "interval_sec", DEFAULT_INTERVAL_SEC,
```

**Review by nokute78 — CHANGES_REQUESTED** [review]
How about supporting `Interval_Sec` and `Interval_Nsec` like other input plugins ?
I think it is good to allow two exclusive settings for consistency and compatibility.
- `Interval_Sec` and `Interval_Nsec`
- `Rate`

Once user sets `Interval_Sec` and/or `Interval_Nsec`, the plugin ignores `Rate` and set polling interval using `Interval_Sec` and `Interval_Nsec`.

File: plugins/in_dummy/in_dummy.c
Comment (franciscovalentecastro) [reply plugins/in_dummy/in_dummy.c]: I agree with the comments and settings proposal. Removed `time_interval` and added `interval_sec/nsec` that overrides the `rate` setting when set.

Comment (franciscovalentecastro) [reply plugins/in_dummy/in_dummy.c]: Re-requesting review from @braydonk and @nokute78.

**Review by braydonk — APPROVED** [review]

**Review by nokute78 — APPROVED** [review]
@franciscovalentecastro Thank you! LGTM.

**Comment by braydonk** [comment]
Those MacOS test failures look unrelated to this PR.

**Comment by braydonk** [comment]
Thanks @nokute78!

`tests/runtime/in_simple_systems.c` doesn't have a listing in CODEOWNERS (would that make sense for it to be you?) 

This means I think we need approval from one of the default folks. @edsiper @leonardo-albertovich if you are able to.

**Review by leonardo-albertovich — APPROVED** [review]

**Pull request merged**
